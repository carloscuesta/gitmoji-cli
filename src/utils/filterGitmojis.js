// @flow
import Fuse from 'fuse.js'

import { type Gitmoji } from '../commands/commit/prompts'

export const options = {
  threshold: 0.5,
  keys: [
    {
      name: 'name',
      weight: 0.33
    },
    {
      name: 'description',
      weight: 0.67
    }
  ]
}

const filterGitmojis = (input: ?string, gitmojis: Array<Gitmoji>) => {
  const fuse = new Fuse(gitmojis, options)

  return input ? fuse.search(input).map((gitmoji) => gitmoji.item) : gitmojis
}

export default filterGitmojis
