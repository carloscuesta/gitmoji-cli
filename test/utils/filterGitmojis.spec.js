import filterGitmojis, { options } from '../../src/utils/filterGitmojis'
import * as stubs from './stubs'

describe('filterGitmojis', () => {
  it('should find all gitmojis with empty input', () => {
    const filteredGitmojis = filterGitmojis(undefined, stubs.gitmojis)

    expect(filteredGitmojis).toStrictEqual(stubs.gitmojis)
  })

  it('should should find the `Updating tests.` gitmoji first', () => {
    const filteredGitmojis = filterGitmojis('test', stubs.gitmojis)

    const gitmoji = stubs.gitmojis.find(
      (gitmoji) => gitmoji.name === 'white-check-mark'
    )
    expect(filteredGitmojis[0]).toStrictEqual(gitmoji)
  })

  it('should should find the `Introducing new features.` gitmoji first', () => {
    const filteredGitmojis = filterGitmojis('fea', stubs.gitmojis)

    const gitmoji = stubs.gitmojis.find(
      (gitmoji) => gitmoji.name === 'sparkles'
    )
    expect(filteredGitmojis[0]).toStrictEqual(gitmoji)
  })

  it('should should find the `Updating code due to external API changes.` gitmoji first', () => {
    const filteredGitmojis = filterGitmojis('alian', stubs.gitmojis)

    const gitmoji = stubs.gitmojis.find((gitmoji) => gitmoji.name === 'alien')
    expect(filteredGitmojis[0]).toStrictEqual(gitmoji)
  })

  it('should match a list of filteredGitmojis', () => {
    const filteredGitmojis = filterGitmojis('bug', stubs.gitmojis)

    expect(filteredGitmojis).toMatchSnapshot()
  })

  it('should not have a total weight of greater than 1', () => {
    const weightSum = options.keys.reduce(
      (carry, key) => (carry += key.weight),
      0
    )

    expect(weightSum).toBeLessThanOrEqual(1)
  })
})
