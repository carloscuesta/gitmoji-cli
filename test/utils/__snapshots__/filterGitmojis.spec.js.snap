// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterGitmojis should match a list of filteredGitmojis 1`] = `
[
  {
    "code": ":bug:",
    "description": "Fixing a bug.",
    "emoji": "🐛",
    "name": "bug",
  },
  {
    "code": ":busts_in_silhouette:",
    "description": "Adding contributor(s).",
    "emoji": "👥",
    "name": "busts-in-silhouette",
  },
  {
    "code": ":bulb:",
    "description": "Documenting source code.",
    "emoji": "💡",
    "name": "bulb",
  },
  {
    "code": ":building_construction:",
    "description": "Making architectural changes.",
    "emoji": "🏗",
    "name": "building-construction",
  },
  {
    "code": ":ambulance:",
    "description": "Critical hotfix.",
    "emoji": "🚑",
    "name": "ambulance",
  },
  {
    "code": ":green_heart:",
    "description": "Fixing CI Build.",
    "emoji": "💚",
    "name": "green-heart",
  },
  {
    "code": ":construction_worker:",
    "description": "Adding CI build system.",
    "emoji": "👷",
    "name": "construction-worker",
  },
]
`;
